doc/admin-guide/plugins/certifier.en.rst

```diff
@@ -0,0 +1,91 @@
+===========
+
+The ``certifier`` performs two basic tasks:
+
+#. Load SSL certificates from file storage on demand. The total number of loaded certificates kept in memory can be configured.
+#. Generates SSL certificates on demand. Generated certificates will be written to file storage for later retrieval.
+
+Implementation
+==============
+Certificates management is done by `SslLRUList` (a Least-Recently-Used (LRU) list with a lookup map). The structure holding all info related to a given SNI is `SslData`. On lookup/insertion, the lookup map will be accessed and the corresponding `SslData` pointer will be moved to the head in the list. If on lookup/insertion the internal count exceeds the given limit on number of files, the tail node will be removed from both the list and the map.
+
+When the plugin sees an incoming HTTPS request, it does:
+
+#. Look up the SNI in `SslLRUList` and set up the context if a valid context exists. Otherwise, it will schedule a thread to retrieve such context from disk (or generate). If such a thread is already scheduled, it will put this SSL connection onto the queue.
+#. The retriever thread will first try to load the cert from disk. If no such certs exist and dynamic generation is enabled, the thread will generate a Certificate Signing Request (CSR) with provided SNI and sign it with root CA passed in from config options. The signed certificate is then written to the disk and all queued up SSL connections are set up with correct context and enabled.
+
+Setup
+=====
+* Cert and key:
+   For example, ``openssl req -newkey rsa:2048 -nodes -keyout ca.key -x509 -days 365 -out ca.cert``. Note that the cert/key should be generated without a challenge password.
+* Serial number:
+   A text file containing a valid integer with a trailing new line.
+
+Plugin Configuration
+====================
+.. program:: certifier.so
+* Specify certificate generation related files. If any of the following parameters is missing, the dynamic generation will be disabled.
+   .. option:: --sign-cert <path_to_certificate>
+   (`optional`, default:empty/unused) - specifies the path to the root CA certficate. In most cases, this would be a self-signed certificate that is configured to be trusted by all potential clients. Path should be the path and file name of the cert. If it is relative, it is relative to the Traffic Server configuration directory.
+
+   .. option:: --sign-key <path_to_key>
+   (`optional`, default:empty/unused) - specifies the path to the root CA private key. In most cases, this would be generated alongside the self-signed root CA certificate.
+
+   .. option:: --sign-serial <path_to_serial>
+   (`optional`, default:empty/unused) - specifies the path to the serial number file. This will be used to assign serial numbers to certificates and keep all generated ones in sync. Serial file should be a number with a trailing newline.
+* Specify the certificates management related settings.
+   .. option:: --store <path_to_certs_dir>
```

Comment:
Is there an upper limit on the number of certs in the directory?  How do they get removed?

Reply:
Right now there is no control over the number of certs in the directory. Not sure if the on-disk certs management should be done in the plugin or elsewhere.